[CL2-6820] Add slug support for folders

nTraum opened the pull request:
The PR introduces slug support for folders. Slugs are friendly URLs that provide a better user experience:

```
Nice slug: /folders/make-the-city-greener
Not so nice UUID: /folders/abcde-12345-fghijk-6789
```

Slug support has existed for projects already. This change introduces it for folders now. It can be edited and will be validated same as projects, so users should feel at home right away. The back-end was ready already to support slugs, we simply have to add the attribute to the JSON API endpoint. :partying_face: 

The translations are now a bit more generic so they work for any kind of resource the slug is used for.

## Checklist

- [x] Added entry to changelog
<details>
<summary>More info</summary>
Add a concise line to the 'Next release' section of the changelog (docs/README.md) so people other than developers can understand what has changed where. E.g. 'Added an error message to the project name field of the project edit form (Admin > Projects > Edit)'.
</details>

- [x] WCAG 2.1 AA proof
<details>
<summary>More info</summary>
For front-end devs only. Is your work conforming with the WCAG 2.1 AA rules? If you need more info, read the [a11y page](https://www.notion.so/citizenlab/a11y-7568f83d42ab4895ac133b89d358997b) on our Notion.
</details>

- [x] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [ ] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [Epic Deployment](**put URL here** or remove)

## How urgent is a code review?

Normal urgency.

nTraum committed:
Move ProjectFolder permitted attributes into policy class

nTraum committed:
Remove sqlite3 as a dev dependency from project folders engine gemspec

It's not used.

nTraum committed:
Merge branch 'master' into CL2-6820_folder_slugs

nTraum committed:
Remove project from SlugInput

nTraum committed:
Merge branch 'master' into CL2-6820_folder_slugs

nTraum committed:
Merge branch 'master' into CL2-6820_folder_slugs

nTraum committed:
WIP add slugs to folders

nTraum committed:
Fix locale and tenant retrieval

nTraum committed:
Validate slug

nTraum committed:
Run prettier to revert my accidental formatting

nTraum committed:
Revert irrelevant formatting changes

nTraum committed:
Remove duplicated translations now that the slug ones have been moved to a separate file

nTraum committed:
Remove duplicated regexError too

Commit:
Translations updated by CI (extract-intl)

brentguf commented:
Hey @nTraum,

> ERROR in app/modules/commercial/project_folders/admin/containers/settings/ProjectFolderForm.tsx:463:11
TS2322: Type 'Error | IAppConfiguration' is not assignable to type 'IAppConfiguration'.
Property 'data' is missing in type 'Error' but required in type 'IAppConfiguration'.
  461 |         </SectionField>
  462 |         <SlugInput
> 463 |           currentTenant={tenant!}
      |           ^^^^^^^^^^^^^
  464 |           locale={locale!}
  465 |           slug={slug}
  466 |           apiErrors={apiErrors}
  
I would personally load appConfiguration (tenant) and locale in the SlugInput component itself with resource hooks. We usually don't pass data from resources as props because it's not necessary; if they've been requested before they're cached and no extra requests are made. So you don't need to worry about passing the props. :)

Second, I would name the SlugInput component. Now it's anonymous (it was already like that). That's not useful in an error stack trace. So I would do

```
const SlugInput = ...

export default injectIntl(SlugInput);
```

instead. 

Also, I would be very wary of using `!`. It should almost never be used, only when TS can't figure it out the type itself for some reason and you've very sure what it's going to be. With `!` we're telling TS to be quiet, which isn't often a good idea.

To finally answer your question: the conditions higher up the page to return null are wrong. The form should not render on these conditions instead:

```
  if (
    // if mode = new we don't have a folder yet, but when we're editing, we need a folder
    (mode === 'edit' && isNilOrError(projectFolder)) ||
    isNilOrError(tenant) ||
    isNilOrError(locale)
  ) {
    return null;
  }
```
 
This will also solve the second error. Will look into the apiErrors problem as well.

brentguf commented:
One other thing: SlugInput should only be rendered when mode === "edit":

{mode === 'edit' && <SlugInput ... />}.

Related to this: the slug prop of SlugInput should actually be string; it shouldn't allow null. So then you would have to do a check to ensure slug is a string (which in this case is equivalent to mode === 'edit' because in 'new' mode, there's no folder and no slug yet. 

In this case, I would do both the mode and slug checks. Strictly, it's not necessary, but we want to be explicit here. Because in theory we could render the SlugInput in 'new' mode as well, but we don't want to at the moment. And as soon as we would allow people to create slugs from scratch, so also when creating a new folder, we would allow null as a value for slug in SlugInput. Does that make sense?

brentguf commented:
For the apiErrors: you can look at how it's done in the file save.ts. In the ProjectFolderForm component, we're dealing with apiErrors in an unorthodox way. So you could add apiErrors to the state (initialize with const [apiErrors, setApiErrors] = useState(...)), then set them in the try/catch block as is done in save.ts. Let me know if you need help here.

brentguf commented:
I unassigned myself. Assign me again after you've solved the errors (otherwise I'll forget). Looks like you're close. :-)

IvaKop commented:
@nTraum On my side, I agree 100% with Brent's remarks above on how to solve the issues you pointed out. I am happy to have another look at this once again when they are resolved + try to answer additional questions, if you have any. We can think about pairing again to add some tests here, as well, if time allows :)

nTraum commented:
Thanks for the comments, super helpful! I'll come back to this PR when I have some time, gonna try to refactor the component on my own and will reach out in case I'm getting stuck.

nTraum committed:
Fetch locale and tenant in SlugInput directly

Commit:
Translations updated by CI (extract-intl)

Commit:
Translations updated by CI (extract-intl)

Commit:
Translations updated by CI (extract-intl)

nTraum committed:
Show SlugInput even when slug is null (empty input)

nTraum committed:
Remove slug null check from FolderForm too

nTraum committed:
Introduce different translations for different wordings

nTraum committed:
prettier formatting

nTraum committed:
Merge branch 'master' into CL2-6820_folder_slugs

nTraum committed:
Remove old SlugInput component

nTraum committed:
Remove unused isNil import

nTraum committed:
Fix slug input error message not validating to true after it has been false once

nTraum committed:
Make SlugInput work for folders

nTraum committed:
Fix style in FoldersController

nTraum committed:
Fix eslint errors

nTraum committed:
Remove dead code

nTraum committed:
Remove unused Input import

nTraum committed:
Merge branch 'master' into CL2-6820_folder_slugs

nTraum committed:
Use more generic translations for slugs so it's not project-specific anymore but can be reused.

Commit:
Translations updated by CI (extract-intl)

nTraum committed:
Fix generated slug preview URL to take into account the correct resource

nTraum committed:
Fix URL preview in tooltip too

nTraum committed:
Add changelog entry

nTraum commented:
@brentguf @IvaKop Could you take another look? I'm happy with the code now.

Epic deployment is in progress, currently [struggling](https://citizenlabco.slack.com/archives/C65GX921W/p1643125582019000) to set it up.

jinjagit commented:
@nTraum @IvaKop @brentguf Epic deployment here: https://en100.epic.citizenlab.co/en/

IvaKop commented on front/app/components/UI/Error/messages.ts:
```diff
@@ -156,7 +156,7 @@ export default defineMessages({
     id: 'app.errors.slug_taken',
     defaultMessage:
-      'This project URL already exists. Please change the project slug to something else.',
+      'This URL already exists. Please change the slug to something else.',
```
Just a heads up, this doesn't affect the translation - should be changed in Crowdin.

nTraum replied:
:+1: How do I go about this? Change the English translation and then our translators will be updating the other languages accordingly on their own?

IvaKop replied:
It has to be done on Crowdin! I usually just leave Karel a message in the Jira ticket and he takes it from there.

IvaKop commented on front/app/components/admin/SlugInput/index.tsx:
```diff
@@ -72,16 +76,16 @@ export default injectIntl(
         <b>{formatMessage(messages.resultingURL)}</b>:{' '}
-        {currentTenant?.data.attributes.host}/{locale}/projects/
-        {slug}
+        {currentTenant?.data.attributes.host}/{locale}/
```
This seems to repeat. Why not create a constant with this value to be able to reuse more easily?

nTraum replied:
Good idea, will do!

IvaKop commented on front/app/modules/commercial/project_folders/admin/containers/settings/ProjectFolderForm.tsx:
```diff
@@ -278,6 +299,8 @@ const ProjectFolderForm = ({ mode, projectFolderId }: Props) => {
           }
+        } catch (errors) {
+          setErrors(errors.json.errors);
         } finally {
```
I am still confused by this finally here... Does it always get called? Feels like `setStatus('apiError');` should be in the catch block.

nTraum replied:
Yes, I've just checked this again locally and applied the proposed change, makes sense!

IvaKop reviewed (approved):
Nice, looks good! 👍 
Congrats on your first React feature! Let me know if/when you want to dive into unit testing as well 🧪

nTraum committed:
Fix apiError always being set even on successfull form submission

nTraum committed:
Remove duplication and move previewUrl into a const

nTraum committed:
Attempt to add a unit test for SlugInput

nTraum committed:
Add cl intl mock

nTraum committed:
styled-component fails

See https://github.com/styled-components/styled-components/issues/1449, but the proposed solution does not work for me.

 FAIL  app/components/admin/SlugInput/SlugInput.test.tsx
  ● Test suite failed to run

    Cannot create styled-component for component: undefined.

      32 | import messages from './messages';
      33 |
    > 34 | const StyledSectionField = styled(SectionField)`
         |                            ^
      35 |   max-width: 100%;
      36 |   margin-bottom: 40px;
      37 | `;

      at options (node_modules/styled-components/src/constants.js:4:33)
      at options (node_modules/styled-components/src/models/StyledComponent.js:271:68)
      at ke (node_modules/styled-components/src/models/StyledComponent.js:271:68)
      at Object.<anonymous> (app/modules/commercial/project_folders/admin/components/ProjectFolderSelect/index.tsx:34:28)
      at Object.<anonymous> (app/modules/commercial/project_folders/index.tsx:8:1)
      at Object.<anonymous> (app/modules/index.ts:3:1)
      at Object.<anonymous> (app/utils/streams.ts:34:1)
      at Object.<anonymous> (app/services/appConfiguration.ts:2:1)
      at Object.<anonymous> (app/services/locale.ts:36:1)
      at Object.<anonymous> (app/resources/GetLocale.tsx:4:1)
      at Object.<anonymous> (app/utils/cl-router/Link.tsx:5:1)
      at Object.<anonymous> (app/components/admin/Section/index.tsx:3:1)
      at Object.<anonymous> (app/components/admin/SlugInput/index.tsx:6:1)
      at Object.<anonymous> (app/components/admin/SlugInput/SlugInput.test.tsx:3:1)
      at TestScheduler.scheduleTests (node_modules/@jest/core/build/TestScheduler.js:333:13)
      at runJest (node_modules/@jest/core/build/runJest.js:404:19)

IvaKop committed:
Fix test

nTraum committed:
Add a real unit test

nTraum committed:
Add a test for failed validation too

nTraum committed:
Remove unused variable assignment

nTraum committed:
Fix lint errors

nTraum committed:
Merge branch 'master' into CL2-6820_folder_slugs

nTraum committed:
Merge branch 'master' into CL2-6820_folder_slugs
